### PR TITLE
Generalize molecular hamiltonian function

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
The PR adds the option for selecting a backend when molecular Hamiltonians are created. The supported options are the local differentiable Hartree-Fock solver and the openfermion-pyscf workflow. 

**Description of the Change:**
The function `molecular_hamiltonian` is modified to have an argument for selecting a differentiable or a non-differentiable  backend. Two different atomic coordinates formats, 1D or 2D arrays, are supported for the user convenience.

**Benefits:**
The UI for building a molecular Hamiltonian becomes more flexible.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/qchem/openfermion_obs.py [hunk 1]
```diff
@@ -911,7 +912,25 @@ def molecular_hamiltonian(
     """
     openfermion, _ = import_of()
 
-    hf_file = meanfield(symbols, coordinates, name, charge, mult, basis, package, outpath)
+    if len(coordinates) == len(symbols) * 3:
+        geometry_dhf = coordinates.reshape(len(symbols), 3)
+        geometry_hf = coordinates
+    elif len(coordinates) == len(symbols):
+        geometry_dhf = coordinates
+        geometry_hf = coordinates.flatten()
```
Comment (soranjh) [inline pennylane/qchem/openfermion_obs.py:920]: These are added to support two atomic coordinates _input_ formats: a flatten 1D array and a 2D array with a Nx3 shape.

File: pennylane/qchem/openfermion_obs.py [hunk 2]
```diff
@@ -911,7 +912,25 @@ def molecular_hamiltonian(
...
+
+    if package == "dhf":
```
[the remark this answers is not in the document]
Comment (ixfoduap) [reply pennylane/qchem/openfermion_obs.py:922]: I would prefer to rename to `method` if it is not too much extra work

[the remark this answers is not in the document]
Comment (ixfoduap) [reply pennylane/qchem/openfermion_obs.py:922]: Also, is `dhf` the default method, as it should be?

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
 import pytest
 
+from pennylane import Identity, PauliX, PauliY, PauliZ
+from pennylane import numpy as np
 from pennylane import qchem
 from pennylane.ops import Hamiltonian
 
 # TODO: Bring pytest skip to relevant tests.
-if not (3, 9) < sys.version_info < (3, 10):
+if not (3, 8) < sys.version_info < (3, 10):
```
Comment (ixfoduap) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:11]: why this change in this PR?

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 4]
```diff
@@ -1,14 +1,14 @@
-import os
 import sys
 
-import numpy as np
 import pytest
 
+from pennylane import Identity, PauliX, PauliY, PauliZ
+from pennylane import numpy as np
```
Comment (ixfoduap) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:6]: why this change in this PR?

**Review by ixfoduap — CHANGES_REQUESTED** [review]
Looks good! It's a really nice UI. Just one major point:
- Let's use `method` instead of `package` and let's make sure `dhf` is the default (wasn't clear to me)

File: pennylane/qchem/openfermion_obs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/openfermion_obs.py:922]: Used `method`.

File: tests/qchem/of_tests/test_molecular_hamiltonian.py (hunk 3 above)
Comment (soranjh) [reply tests/qchem/of_tests/test_molecular_hamiltonian.py:11]: Corrected.

File: tests/qchem/of_tests/test_molecular_hamiltonian.py (hunk 4 above)
Comment (soranjh) [reply tests/qchem/of_tests/test_molecular_hamiltonian.py:6]: We need pennylane numpy to make sure we can use `requires_grad` when building a Hamiltonian with `'dhf'` and the Pauli operators are needed to build the pennyane Hamiltonian. The original `test_molecular_hamiltonian` was only asserting: 
```
assert isinstance(built_hamiltonian, Hamiltonian)
assert qubits == 2 * nact_orbs
```

File: pennylane/qchem/openfermion_obs.py [hunk 5]
```diff
@@ -812,19 +812,23 @@ def molecular_hamiltonian(
     charge=0,
     mult=1,
     basis="sto-3g",
-    package="pyscf",
+    method="pyscf",
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:815]: Do you think that we default the method to `dhf`? 🤔

**Review by obliviateandsurrender — APPROVED** [review]
Thanks @soranjh! Looks good. Just left one small query.

**Review by ixfoduap — APPROVED** [review]

File: pennylane/qchem/openfermion_obs.py (hunk 5 above)
Comment (soranjh) [reply pennylane/qchem/openfermion_obs.py]: It is the default.
